Drop route service namespace

liggitt opened the pull request:
Fixes #7204

deads2k commented on pkg/route/registry/route/strategy.go:
```diff
@@ -43,6 +43,8 @@ func (routeStrategy) NamespaceScoped() bool {
 	route := obj.(*api.Route)
 	route.Status = api.RouteStatus{}
+	// Do not allow cross-namespace routes
+	route.Spec.To.Namespace = ""
```
Doesn't this scream `LocalObjectReference` to you?

liggitt replied:
LocalObjectReference doesn't have a kind, which I assume we'd want to be able to set eventually... could make it one for now, and expand later?

deads2k replied:
> LocalObjectReference doesn't have a kind, which I assume we'd want to be able to set eventually... could make it one for now, and expand later?

Let's add a kind.  That ought to be pretty reasonable upstream right?

liggitt replied:
I'll do that, but that'll be queued behind the codecgen changes I have. I'd rather get this in now and drop it when I convert to a LocalObjectReference

deads2k replied:
> I'll do that, but that'll be queued behind the codecgen changes I have. I'd rather get this in now and drop it when I convert to a LocalObjectReference

Ok.  Open an issue please.

deads2k commented:
lgtm pending issue to change the `ObjectReference` to a `LocalObjectReference`

liggitt commented on pkg/route/api/v1/defaults.go:
```diff
@@ -0,0 +1,29 @@
+func addDefaultingFuncs(scheme *runtime.Scheme) {
+	err := scheme.AddDefaultingFuncs(
+		func(obj *RouteSpec) {
+			if len(obj.To.Kind) == 0 {
```
defaulting is conditional now

liggitt commented on pkg/route/api/v1beta3/defaults.go:
```diff
@@ -0,0 +1,29 @@
+func addDefaultingFuncs(scheme *runtime.Scheme) {
+	err := scheme.AddDefaultingFuncs(
+		func(obj *RouteSpec) {
+			if len(obj.To.Kind) == 0 {
```
same here

liggitt commented:
[test]

liggitt commented:
Opened https://github.com/openshift/origin/issues/7370

Also added validation for kind, and updated defaulting to only set kind to Service if empty, so we fail validation if values other than "Service" are sent to the API, rather than silently overwriting.

liggitt commented:
@deads2k still look good to you? These bits changed after your review:
https://github.com/openshift/origin/pull/7212/files#diff-1152bfa99c0917bf00854302f786a2ebR43
https://github.com/openshift/origin/pull/7212/files#diff-13cdc93be4a264426a1f7c00583e996dR48

deads2k commented:
lgtm [merge]
